Adds size variable at runtime for arrays that cannot be inferred

njriasan opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities here, please click the arrow besides
   "Create Pull Request" and choose "Create Draft Pull Request".
   When it's ready for review, you can click the button "ready to review" near
   the end of the pull request
   (besides "This pull request is still a work in progress".)
   The maintainers will then be automatically notified to review it.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on main/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against main they should
   be resolved by merging main into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

Closes #8146

Skips replacing an array expression with a parfor if due to control flow the size var could not be found in the array. This doesn't enable parallelism but it does allow the code to compile and run. Its seems like we may want output a warning in this situation.

DrTodd13 reviewed (changes requested):
The root cause of this problem should be fixed which is to make sure that the size variable exists.  If an expression needs a size variable but it doesn't exist then we should emit code to generate the size var at that point.  I don't want to set the precedent that we just don't parallelize things due to a missing size var.  We already have a routine I believe for generating size vars for an array so why can't you just iterate through the ``size_vars`` and if one is not available then generate it at that point in the code using the routine I mentioned?

njriasan commented:
@DrTodd13 Agree 100% with what you are saying here. Do these changes accomplish your request?

DrTodd13 commented:
> @DrTodd13 Agree 100% with what you are saying here. Do these changes accomplish your request?

Thanks Nick!  I really appreciate the change and I think that the code inside the "elif" is likely correct.  What isn't clear to me is whether the "elif" conditional is exactly right and whether similar changes need to be done elsewhere.  I don't expect you to cover all such cases but what I would like to see at this point are a few additional tests.  1) Rather than divide by a scalar, have a test where you do element-wise division by another array where the second array is known to be a certain size.  2) Another test similar to the first except that the second array's size is similarly unknown at that point just like the first.  My expectation is that you are going to have some arrays in the broadcast that you know the size of and others you don't and that your code needs to be run for each array that you don't know the size.  So, for the first new test you'll only need to run it on one array but it won't always be the first one (so len(arrs) == 1 won't be true).  For the second new test, you'll have to run this code for both of the input arrays.

njriasan commented:
@DrTodd13 I added the additional tests and you are correct there are still some issues that exist. In particular, there appear to be issues with `_insert_runtime_broadcast_call`. I fixed compilation errors in `runtime_broadcast_assert_shapes`, but I believe this code should produce an assert somewhere, but it doesn't.

In addition, I believe this output shape be "equivalent" to the lhs of the array expression as we verify this at runtime. However, I'm not sure the proper way to do this. Any advice would be appreciated.

njriasan commented:
I added an example of what I think the assert should be, but I'm not sure its correct.

njriasan commented on numba/parfors/array_analysis.py:
```diff
@@ -2997,9 +3006,21 @@ def _insert_runtime_broadcast_call(self, scope, loc, arrs, max_dim):
+        # Define the output in the equiv_set for the left hand side.
+        # These should be considered equivalent because we produced a
+        # runtime check that verifies this.
+        needs_define = equiv_set.insert_equiv(runtime_broadcast_shape, *arrs)
```
I don't think this is correct, but I need a way to ensure the `runtime_broadcast_shape` is selected by the lhs of array_expr. This gets `test_2array_1_control_flow`, but again I don't think its correct.

njriasan commented:
@DrTodd13 discussed this with me on a call and he suggested that rather than have the separate paths for different shapes that we generate the shapes when needed for any that are previously unknown and then call `_broadcast_assert_shapes`. Making this change passes all of my tests and allows me to eliminate some unused code.

DrTodd13 reviewed (approved):
Looks good to me and all the tests seem to pass.

njriasan commented:
@guilhermeleobas  For parfor changes does this need a second review or since Todd has approved can we merge this PR?

adofarsi commented:
Dear @DrTodd13,
I have tested @njriasan 's dummy reproducer code in #8146  and it still fails with current release of Numba (0.56.4) on both Mac (macOS-13.0.1-x86_64-i386-64bit) and Linux VM (Linux-5.15.0-1018-gcp-x86_64-with-glibc2.35) with output:

```
func_fnty: Function(<function runtime_broadcast_assert_shapes at 0x116ee3040>)
typs: [array(int64, 1d, C)]
dims: [1] 1
fargs: [Literal[int](1), array(int64, 1d, C)]
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
/var/folders/bh/g86lgxvj67x8_qdx30vwmw500000gn/T/ipykernel_44644/293618984.py in <module>
     14 
     15 arr = np.arange(100)
---> 16 print(f(arr, True, False))

~/opt/anaconda3/envs/ado/lib/python3.9/site-packages/numba/core/dispatcher.py in _compile_for_args(self, *args, **kws)
    466                 e.patch_message(msg)
    467 
--> 468             error_rewrite(e, 'typing')
    469         except errors.UnsupportedError as e:
    470             # Something unsupported is present in the user code, add help info

~/opt/anaconda3/envs/ado/lib/python3.9/site-packages/numba/core/dispatcher.py in error_rewrite(e, issue_type)
    407                 raise e
    408             else:
--> 409                 raise e.with_traceback(None)
    410 
    411         argtypes = []

TypingError: Failed in nopython mode pipeline (step: convert to parfors)
No implementation of function Function(<function runtime_broadcast_assert_shapes at 0x116ee3040>) found for signature:
 
 >>> runtime_broadcast_assert_shapes(Literal[int](1), array(int64, 1d, C))
 
There are 2 candidate implementations:
 - Of which 2 did not match due to:
 Overload in function 'register_jitable.<locals>.wrap.<locals>.ov_wrap': File: numba/core/extending.py: Line 159.
   With argument(s): '(int64, array(int64, 1d, C))':
  Rejected as the implementation raised a specific error:
    TypeError: missing a required argument: 'arg1'
  raised from /Users/adofarsi/opt/anaconda3/envs/ado/lib/python3.9/inspect.py:2960
```

gmarkall commented:
@adofarsi This PR is not part of Numba 0.56.4 - were you expecting some different result to the one you observed?

adofarsi commented:
Hi @gmarkall, thanks for the prompt reply.

Yes. For example the dummy reproducer code in https://github.com/numba/numba/issues/8146 works fine with flag `parallel=False`:
```
import numpy as np
import numba

@numba.njit(parallel=False)
def f(arr, flag1, flag2):  # pragma: no cover
    inv = np.arange(arr.size)
    if flag1:
        return inv.astype(np.float64)
    if flag2:
        ret = inv[inv]
    else:
        ret = inv[inv - 1]
    return ret / arr.size

arr = np.arange(100)
print(f(arr, True, False))
```
and returns the following expected output:
```
[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34. 35.
 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52. 53.
 54. 55. 56. 57. 58. 59. 60. 61. 62. 63. 64. 65. 66. 67. 68. 69. 70. 71.
 72. 73. 74. 75. 76. 77. 78. 79. 80. 81. 82. 83. 84. 85. 86. 87. 88. 89.
 90. 91. 92. 93. 94. 95. 96. 97. 98. 99.]
```
I have a similar issue in my code where a jitted function works fine with flag `parallel=False` but then outputs the same `function runtime_broadcast_assert_shapes` error when jitted with flag `parallel=True`.

gmarkall commented:
I would not expect any issues resolved by this PR to be resolved in Numba 0.56.4. Can you clarify what you are expecting and why, please?

adofarsi commented:
Do you know when this will be included in a new Numba release? And would you be able to suggest any workaround around this problem until then? Thank you for your time.

gmarkall commented:
It will be included in the 0.57 release, which will be early next year. As a workaround for the meantime, you could install Numba from the Numba dev channel (if you're using conda) or build it from source.

- To install from the Numba dev channel with conda, use `conda install numba/label/dev::numba`
- To install from source, see: https://numba.readthedocs.io/en/latest/user/installing.html#installing-from-source

adofarsi commented:
Hi again,

Thank you for your reply. I have reinstalled Numba from the Numba dev channel using `conda install numba/label/dev::numba`, verified that the correct package was installed with `conda list | grep numba` and output:
```
numba                     0.56.4          np1.16py3.9h2fabc88_g288a38bbd_0    numba/label/dev
```
I have then rerun the code in https://github.com/numba/numba/issues/8146 and it still fails with the following output:
```
func_fnty: Function(<function runtime_broadcast_assert_shapes at 0x114244940>)
typs: [array(int64, 1d, C)]
dims: [1] 1
fargs: [Literal[int](1), array(int64, 1d, C)]
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
/var/folders/bh/g86lgxvj67x8_qdx30vwmw500000gn/T/ipykernel_98678/293618984.py in <module>
     14 
     15 arr = np.arange(100)
---> 16 print(f(arr, True, False))

~/opt/anaconda3/envs/ado/lib/python3.9/site-packages/numba/core/dispatcher.py in _compile_for_args(self, *args, **kws)
    466                 e.patch_message(msg)
    467 
--> 468             error_rewrite(e, 'typing')
    469         except errors.UnsupportedError as e:
    470             # Something unsupported is present in the user code, add help info

~/opt/anaconda3/envs/ado/lib/python3.9/site-packages/numba/core/dispatcher.py in error_rewrite(e, issue_type)
    407                 raise e
    408             else:
--> 409                 raise e.with_traceback(None)
    410 
    411         argtypes = []

TypingError: Failed in nopython mode pipeline (step: convert to parfors)
No implementation of function Function(<function runtime_broadcast_assert_shapes at 0x114244940>) found for signature:
 
 >>> runtime_broadcast_assert_shapes(Literal[int](1), array(int64, 1d, C))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'register_jitable.<locals>.wrap.<locals>.ov_wrap': File: numba/core/extending.py: Line 159.
    With argument(s): '(int64, array(int64, 1d, C))':
   Rejected as the implementation raised a specific error:
     TypeError: missing a required argument: 'arg1'
  raised from /Users/adofarsi/opt/anaconda3/envs/ado/lib/python3.9/inspect.py:2960
```

gmarkall commented:
Something seems to have held Numba back at 0.56.4 and installed that again, just from the dev channel this time. If you do `conda upgrade -c numba/label/dev numba` does it install 0.57.0dev0...` instead?
